py/testdir_multi_jvm/test_gbm_prostate.py

```diff
@@ -21,7 +21,7 @@
 # Connect to a pre-existing cluster
 h2o.init(ip=ip, port=port, strict_version_check=False)
 
-df = h2o.import_file(path="../../smalldata/logreg/prostate.csv")
+df = h2o.import_file(path=os.path.realpath("../../smalldata/logreg/prostate.csv"))
```

Comment:
This illustrates this change could be a breaking change for some users who are relying on current behavior.

This is most likely the only test that uses h2o.import file directly, the other tests are probably using locate:

    df = h2o.import_file(path=pyunit_utils.locate("smalldata/logreg/prostate.csv"))

Can we detect that the client is running on the same machine as the h2o backend and revert to the original behavior in that case? Eg. by writing a temporary file and trying to read it from H2O while verifying the content?

We can also have a configuration option that will switch between old and new behaviors.

WDYT?

Reply:
@michalkurka : we can't have it both ways:
* we use R impl. as the reference, and we break this support for relative local path that existsonly for the Python lib. R impl doesn't do any path manipulation at all.
* we keep this Python support to avoid breaking changes: that's what my original changes were trying to achieve at a cost of additional complexity, and in my opinion, creating a misunderstanding when the cluster is not running on the local machine, as the user thinks this path parameter is a local path instead of a path provided to the backend in order to load the file.

I think switching between behaviours is overhead and will add more confusion for us in the future when people will report import issues.

Reply:
if we want to be safe and not introduce breaking changes this time, I'd suggest to use my original fix (with correct expansion of relative paths into file URI on all OS) for this release and create a follow-up ticket for the next stable release, where we can clean-up this fix and replace it with simple alignment with R impl.

Reply:
@seb-h2o It makes sense to avoid additional complexity. 

Most of our community users/enthusiast is running the client on the same machine where the backend is running. The existing behavior of the python client is to expand the path on the client. This bug has been there for over 2 years and in my opinion, we cannot simply defer to the documentation that says we are doing server side expansion.

I would not change the behavior in a fix release. My solution would be to introduce an option that would let the user switch between local and remote expansion. Default would be local (==as is). Users who are reporting the issue can simply flip the switch to remote expansion.

Reply:
ok.. although I'm afraid we're promoting an old bug to a pseudo-feature here: typically, `h2o.import_file(path="./path/to/file")` is behaving differently regarding path expansion than `h2o.import_file(path=["./path/to/file1", "./path/to/file2"])`, showing how broken it is.

I'll fix this with additional param on the python api, and will create follow-up ticket to ensure this get removed in next stable release.

Reply:
another idea to educate user: remove the hack but have an assertion instead: if provided path is relative, let the api behave normally, but if error is raised, also print a warning message, something like "please ensure that the given path is accessible from the server, even if you're running the server locally, prefer providing an absolute path"

Reply:
Agreed, it is not ideal. But we should really not break stuff for users in the fix release cycle. We should clean this up in a major release.

I would not introduce a parameter, more like a flag similar to let's say this one: https://github.com/h2oai/h2o-3/blob/master/h2o-py/h2o/expr.py#L80

Affected users can flip this switch on. We should document that this is a temporary measure.

Reply:
@seb-h2o that is a good idea but it only works in an interactive mode. My concern is that we will break someones unsupervised production scripts.

Reply:
good point